sw/device/lib/otbn.h

```diff
@@ -0,0 +1,248 @@
+  {                                                   \
+    .imem_start = _otbn_app_##app_name##__imem_start, \
+    .imem_end = _otbn_app_##app_name##__imem_end,     \
+    .dmem_start = _otbn_app_##app_name##__dmem_start, \
+    .dmem_end = _otbn_app_##app_name##__dmem_end,     \
+  }
+
+/**
+ * Make a symbol in the OTBN application image available.
+ *
+ * Symbols are typically function or data pointers, i.e. labels in assembly
+ * code.
+ *
+ * @param app_name    name of the application the function is contained in
+ * @param symbol_name name of the symbol (function, label)
+ */
+#define OTBN_USE_PTR(app_name, symbol_name) \
+  extern const char _otbn_app_##app_name##_##symbol_name[];
+
+/**
+ * Initialize an `otbn_ptr_t` structure.
+ */
+#define OTBN_PTR_T_INIT(app_name, symbol_name) \
+  { .ptr_in_cpu_addr_space = _otbn_app_##app_name##_##symbol_name, }
+
+/**
+ * Initialize the OTBN context structure.
+ *
+ * @param ctx the context object
+ * @param dif_config the OTBN DIF configuration (passed on to the DIF)
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnBadArg if a precondition check failed,
+ *         #kOtbnError otherwise
+ */
+otbn_result_t otbn_init(otbn_t *ctx, const dif_otbn_config_t *dif_config);
+
+/**
+ * (Re-)load the RSA application into OTBN.
+ *
+ * Load the application image with both instruction and data segments into OTBN.
+ *
+ * @param ctx the context object
+ * @param app the application to load into OTBN
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnBadArg if a precondition check failed,
+ *         #kOtbnError otherwise.
+ */
+otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t *app);
+
+/**
+ * Call a function on OTBN.
+ *
+ * Set the entry point (start address) of OTBN to the desired function, and
+ * starts the OTBN operation.
+ *
+ * Use `otbn_busy_wait_for_done()` to wait for the function call to complete.
+ *
+ * @param ctx the context object
+ * @param func the function to be called
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnFuncNotInApp if `func` isn't part of the loaded application,
+ *         #kOtbnError otherwise.
+ */
+otbn_result_t otbn_call_function(otbn_t *ctx, const otbn_ptr_t *func);
+
+/**
+ * Busy wait for OTBN to be done with its operation.
+ *
+ * @param ctx the context object
+ * @return #kOtbnOk if the operation was successful,
+ *         #kOtbnExecutionFailed if the application execution failed,
+ *         #kOtbnError otherwise.
+ */
+otbn_result_t otbn_busy_wait_for_done(otbn_t *ctx);
+
+/**
+ * Copy data from the CPU memory to OTBN data memory.
+ *
+ * @param ctx the context object
+ * @param dest pointer to the destination in OTBN's data memory
+ * @param src source of the data to copy
+ * @param len_bytes number of bytes to copy
+ * @return #kOtbnOk if the transfer was successful,
+ *         #kOtbnBadArg precondition checks failed,
+ *         #kOtbnError otherwise.
```

Comment:
This type of @return isn't super userful, since it repeats something that is already part of result_t's contract. Consider abridging it?